Openthread (#103)
AVSystem/developer-zone: docs/LwM2M_Client/OpenThread/Adding_device.md

# Adding nRF board to OpenThread

Integrate your Nordic nRF52840-DK board to an active OpenThread Border Router.

## Prerequisites

- An active **OpenThread Border Router**.
- The **nRF52840-DK** board with a USB cable.
- Installed **minicom** or **RealTerm** (for Linux or Mac) or **PuTTy** (for Windows) or other serial communication program.
- Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
- A user with access to the [**Coiote IoT Device Management**](https://eu.iot.avsystem.cloud/ui/device/inventory) platform and appropriate permissions.

## Connecting to the OpenThread

Check your active Border Router **OTBR_IP_address** and **OTBR_port** on which your Border Router works. This will be necessary for further steps.

### Connection through Wi-Fi

> You have to be connected to the same Wifi network as your OpenThread Border Router. To connect your device to Border Router through Wifi, you should go to `OTBR_IP_address`.

Reviewer: It's not clear what you mean by "go to `OTBR_IP_address`. Maybe state that you can connect to the WebUI using a web browser by connecting to `OTBR_IP_address:OTBR_port`

!!! Note

>     There is an option to connect to your Border Router through ssh by `ssh -L OTBR_port:OTBR_IP_addres:80 -L 8081:OTBR_hostname` command, then go to: `http://OTBR_IP_address:80`. The additional forwarding for port 8081 is necessary for the ***Topology*** view to work. Also, port 80 is necessary due to the HTTP protocol.

Reviewer: I'm not sure what you wanted to to say here. The first part of the part of the `ssh` command `-L OTBR_port:OTBR_IP_addres:80` means port forwarding. You forward your OTBR_port (on your local machine) to OTBR_IP_addres:80 but your not connecting to anything with the ssh command so I guess you won't forward anything. We are using port forwarding to connect to our OTBR but this may not be the case for someone outside our team.

When you are connected you should see the OpenThread main page:

![OT Main Page](images/ot_main_page.PNG "OT Main Page")

### Creating an OpenThread network

> On the left side, select the option ***Form***, and a new page will be displayed for the network creation. Click the form and a message should pop up to let you know about the operation's success. ![Form Network](images/form_network.png "Form Network"){:style="float: left;margin-right: 1177px;margin-top: 17px; margin-bottom: 17px;"}

Reviewer: Did you check if the default configuration from this ***Form*** page works?

Author: I read about this [here](https://jerome-silabs.github.io/SE_FAE_team/OpenThread/Applications/OpenThread_Border_Router/create_network.html) but I didn't try this

On the left-side menu, click ***Topology*** to see the role of Border Router ![Border Router topology](images/border1.png "Border Router topology"){:style="float: left;margin-right: 1177px;margin-top: 17px; margin-bottom: 17px;"}

> ### Starting Commission

Reviewer: Maybe "commissioning a device" or something like that?

To start the commissioning process click ***Commission*** on the left-side menu and write down Joiner PSKd password.

!!! Important

>     The PSKd needs to conform to the following: Length: 6-32 characters, encoding: base32-thread (0-9, A-Y excluding I, O, Q, and Z for readability).

Reviewer: OK this page that you link https://community.silabs.com/s/article/openthread-faq?language=en_US does refers to this as a base32-thread encoding. But this one does not: https://openthread.io/guides/border-router/external-commissioning/prepare?hl=en (sating just that it should be 0-9, A-Y excluding I, O, Q, and Z for readability). I did tried to look up what base32-thread encoding means and found https://github.com/openthread/openthread/issues/2156 where someone asks if there is a format reference for this. And also a link to a different link: https://github.com/openthread/openthread/pull/2185 that I didn't read through. Maybe if we wand to use the phrase base32-thread that we should find the official reference and link that.

    <br />
    For more information check [here](https://community.silabs.com/s/article/openthread-faq?language=en_US).

Click start commission, and a message should pop up to inform you about the operation's success. ![Start commission](images/commision.png "Start commission"){:style="float: left;margin-right: 1177px;margin-top: 17px; margin-bottom: 17px;"}

### Device configuration

0. Connect the nRF52840 board to the USB port of your machine.

0. Because NCS uses a different Zephyr version, it is necessary to change our Zephyr workspace, it is handled by using a different manifest file.
Set West manifest path to *Anjay-zephyr-client/demo*, and manifest file to *west-nrf.yml* and do *west update*.

    ```
        west config manifest.path Anjay-zephyr-client/demo
        west config manifest.file west-nrf.yml
        west update
    ```

0. Go to *Anjay-zephyr-client/demo/boards* folder and find `nrf52840dk_nrf52840.conf` file. In this file, you will need to change the joiner PSKd password (line 41 in the .conf file). ![Configuration file](images/conf_file.PNG){:style="float: left;margin-right: 1177px;margin-top: 17px; margin-bottom: 17px;"}

    !!! Note

>         The last config option in file `CONFIG_OPENTHREAD_FTD` tells about a Full Thread Device (FTD) which always has its radio on and maintains IPv6 address mappings. This option can be changed to `CONFIG_OPENTHREAD_MTD`, a Minimal Thread Device (MTD) forwards all messages to its Parent.

Reviewer 2: ```suggestion
        The last config option in the `nrf52840dk_nrf52840.conf` file (`CONFIG_OPENTHREAD_FTD`) tells about a Full Thread Device (FTD) which always has its radio on and maintains IPv6 address mappings. This option can be changed to `CONFIG_OPENTHREAD_MTD`, a Minimal Thread Device (MTD) that forwards all messages to its Parent.
```

> 0. Now in demo directory you can compile the project.

Reviewer 2: ```suggestion
0. Now in _Anjay-zephyr-client/demo_ directory you can compile the project.
```

    ```
        west build -b nrf52840dk_nrf52840
    ```

0. After successful build you can flash the target.

    ```
        west flash
    ```

## Connecting to the LwM2M Server

To connect to Coiote IoT Device Management LwM2M Server, please register at [https://eu.iot.avsystem.cloud](https://eu.iot.avsystem.cloud).

To connect the board:

1. Log in to Coiote DM and from the left side menu, select **Device Inventory**.
2. In **Device Inventory**, click **Add device**.
3. Select the **Connect your LwM2M device directly via the Management server** tile.
       ![Add via Mgmt](images/mgmt_tile.png "Add via Mgmt")
    3. In the **Device credentials** step:

>          - In the **Device ID** enter your board endpoint name, e.g. `test_device`.

Author: This is a copy-and-paste part from other tutorials. If we want to change this here then we need to change this everywhere. @Mierunski

>          - In the **Device ID** enter your board endpoint name, e.g. `test_device`.

Reviewer: OK, I won't enforce this if it's already in other places, but maybe we will want to change this under some other ticket in the future.

             ![Device credentials step](images/add_mgmt_quick.png "Device credentials step")
         - In the **Security mode** section, select the **PSK (Pre-Shared Key)** mode:
              - In the **Key identity** field, type the same name as in the `Endpoint name` field
              - In the **Key** field, type the shared secret used in the device-server authentication.
    4. Click the **Add device** button and **Confirm** in the confirmation pop-up.
    5. In the **Connect your device** step, follow the [next section](#configuring-the-client) to run the client and connect it to the server.

> ## Configuring the Client

Reviewer: Don't he have some kind of generic tutorial that you could just link saying that here is how to configure the device instead of writing it here?

Author: I think we don't have this. Maybe we can create something like this but I think only few tutorials use this

> 0. With the board still connected to a serial port interface, open a serial communication program.

Reviewer 2: ```suggestion
0. With the board still connected to a serial port interface, open a serial communication program (Zephyr logs should appear on the screen). 
```
or something.

Author: It's copy and paste part from the other tutorial. If we want to change this here we should change this everywhere.

0. Use the `anjay` command to list possible options:

    ```
    uart:~$ anjay
    anjay - Anjay commands
    Subcommands:
    start   :Save config and start Anjay
    stop    :Stop Anjay
    config  :Configure Anjay params
    ```

    !!! tip

>         To show available subcommands, use the **Tab** key.

Author: I used **Tab** with this device and it worked. Also, I checked this and it still works xd

>         To show available subcommands, use the **Tab** key.

Reviewer: Great! I wasn't sure if it will work with default settings.

0. Check your default credentials by following the instructions in the program:
    ![Anjay configuration](images/anjay_config.png "Anjay configuration"){:style="float: left;margin-right: 1177px; margin-top: 7px; margin-bottom: 17px;"}

    !!! note
        If your default credentials are different from device credentials provided in Coiote DM, change them using the `anjay config set <possible_option> <value>` command.
        <br/>
        ![Anjay set configuration](images/anjay_config_set.PNG "Anjay set configuration"){:style="float: left;margin-right: 1177px;margin-top: 7px; margin-bottom: 17px;"}

> 0. Use the `anjay start` command to run the Client.

Reviewer 2: Well, you should probably do it in order to change any anjay's settings. It should be included in the instruction in my opinion.

0. In logs, you can read view lines that contain OpenThread informations. There will be information about success or failure in joining the connection and the current role of your device in the OpenThread network.

    ![Connected as a child role](images/connected_child.PNG "Connected as a child role"){:style="float: left;margin-right: 1177px;margin-top: 7px; margin-bottom: 17px;"}

0. Go to OpenThread main web page, on the left-side menu click ***Topology*** to see connected device.

    ![Connected device as a child](images/border2.png "Connected device as a child"){:style="float: left;margin-right: 1177px;margin-top: 7px; margin-bottom: 17px;"}

    !!! Note

>         The node should join the OTBR Thread network automatically. Within two minutes its state should be `router`

Reviewer: ```suggestion
        The node should join the OTBR Thread network automatically. Within two minutes its state should change to `router`
```

Reviewer 2: Also, add `.` at the end.

# Adding nRF board to OpenThread

Integrate your Nordic nRF52840-DK board to an active OpenThread Border Router.

## Prerequisites

- An active **OpenThread Border Router**.
- The **nRF52840-DK** board with a USB cable.
- Installed **minicom** or **RealTerm** (for Linux or Mac) or **PuTTy** (for Windows) or other serial communication program.
- Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
- A user with access to the [**Coiote IoT Device Management**](https://eu.iot.avsystem.cloud/ui/device/inventory) platform and appropriate permissions.

## Connecting to the OpenThread

Check your active Border Router **OTBR_IP_address** and **OTBR_port** on which your Border Router works. This will be necessary for further steps.

### Connection through Wi-Fi

> You have to be connected to the same Wifi network as your OpenThread Border Router. To connect your device to Border Router through Wifi, open any web browser and connect to the WebUI by connecting to **OTBR_IP_address:OTBR_port**.

Reviewer: There are sure a lot of **connecting** done here. Please try to rewrite this.

# Adding nRF board to OpenThread

Integrate your Nordic nRF52840-DK board to an active OpenThread Border Router.

## Prerequisites

- An active **OpenThread Border Router**.
- The **nRF52840-DK** board with a USB cable.
- Installed **minicom** or **RealTerm** (for Linux or Mac) or **PuTTy** (for Windows) or other serial communication program.
- Installed **nrfjprog** from [Nordic Semiconductor page](https://www.nordicsemi.com/Products/Development-tools/nrf-command-line-tools/download)
- A user with access to the [**Coiote IoT Device Management**](https://eu.iot.avsystem.cloud/ui/device/inventory) platform and appropriate permissions.

## Connecting to the OpenThread

Check your active Border Router **OTBR_IP_address** on which your Border Router works. This will be necessary for further steps.

> ### Connection through Wi-Fi
> 
> You have to be attached to the same Wifi network as your OpenThread Border Router. To join your device to Border Router, open any web browser and connect to the WebUI by typing **OTBR_IP_address:80**.

Reviewer: ```suggestion
### Connecting to OpenThread Boarder Router WebUI

You have to be attached to the same network as your OpenThread Border Router. To configure the Border Router open any web browser and connect to the WebUI by typing **OTBR_IP_address:80**.
```

Author: We've got ## Connecting to the OpenThread two lines above so I don't think it will be good to repeat the title. MAybe something like ###Conneting to WebUI?
